qabelbox/src/main/java/de/qabel/qabelbox/activities/CreateIdentityActivity.kt

```diff
@@ -0,0 +1,257 @@
+import org.json.JSONException
+import org.json.JSONObject
+import javax.inject.Inject
+
+class CreateIdentityActivity : BaseWizardActivity(), QabelLog {
+
+    private val TAG = this.javaClass.simpleName
+    private val REQUEST_READ_PHONE_STATE = 1
+    private var READ_PHONE_STATE_DENIED = false
+
+    private var identityName: String = ""
+    private var phoneNumber: String = ""
+    private var email: String = ""
+
+    var createdIdentity: Identity? = null
+    private var prefix: String? = null
+    internal var tryCount = 0
+
+    @Inject
+    lateinit internal var identityUseCase: IdentityUseCase
+    lateinit internal var dropUrlGenerator: DropUrlGenerator
+
+    private var existingIdentities: Identities? = null
+
+    override val headerFragmentText: String get() = identityName
+
+    override val headerSecondLine: String
+        get() = email
+    override val headerThirdLine: String
+        get() = phoneNumber
+
+    override val actionBarTitle: Int = R.string.headline_add_identity
+
+    override val wizardEntityLabel: String get() = getString(R.string.identity)
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        QabelBoxApplication.getApplicationComponent(applicationContext).inject(this)
+        super.onCreate(savedInstanceState)
+        dropUrlGenerator = DropUrlGenerator(getString(R.string.dropServer))
+        identityUseCase.getIdentities().subscribe({
+            existingIdentities = it
+            canExit = it.identities.size > 0
+        }, {
+            existingIdentities = Identities()
+        })
+    }
+
+    private fun tryReadPhoneNumber() {
+        val permissionCheck = ContextCompat.checkSelfPermission(this, Manifest.permission.READ_PHONE_STATE)
+        if (permissionCheck != PackageManager.PERMISSION_GRANTED) {
+            if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.READ_PHONE_STATE)) {
+                alert(R.string.dialog_headline_info, R.string.phone_number_request_info, {
+                    positiveButton(R.string.yes) { tryReadPhoneNumber() }
+                    negativeButton(R.string.no) { READ_PHONE_STATE_DENIED = true }
+                })
+            } else {
+                ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.READ_PHONE_STATE), REQUEST_READ_PHONE_STATE)
+            }
+        } else {
+            val phoneManager = ctx.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
+            var phone = phoneManager.line1Number
+            if (phone != null && !phone.isEmpty()) {
+                try {
+                    phone = formatPhoneNumber(phone)
+                    (fragments[3] as CreateIdentityEditTextFragment).setValue(phone)
+                    info("Phone number detected $phoneNumber")
+                } catch (ex: NumberFormatException) {
+                    error("Error parsing received system phone number $phone", ex)
+                }
+            }
+        }
```

Comment:
Most of the code is not about reading the phone number but about possible failure conditions. Don't you like guard clauses?
Another approach from cc would be to push the readPhoneNumber to it's own method:
```KOTLIN
    private fun tryReadPhoneNumber() {
        if (isPermissionGranted(this, READ_PHONE_STATE)) {
            readPhoneNumber()
        } else {
            handleMissingPhonePermission()
        }
    }
```
This way you get a clean readPhoneNumber() method and encapsulated failure handling.
But that's more a trade-off concerning the negative conditional.